File: services/ingest/internal/storage/storage.go
package storage

import (
	"context"
	"io"

	"github.com/minio/minio-go/v7"
	"github.com/minio/minio-go/v7/pkg/credentials"
)

type MinioStore struct {
	client *minio.Client
	bucket string
}

func NewMinioStore(endpoint, accessKey, secretKey, bucket string) (*MinioStore, error) {
	cli, err := minio.New(endpoint, &minio.Options{
		Creds:  credentials.NewStaticV4(accessKey, secretKey, ""),
		Secure: false,
	})
	if err != nil {
		return nil, err
	}
	return &MinioStore{client: cli, bucket: bucket}, nil
}

func (s *MinioStore) Put(ctx context.Context, key string, src io.Reader, size int64, contentType string) (int64, error) {
	info, err := s.client.PutObject(ctx, s.bucket, key, src, size, minio.PutObjectOptions{
		ContentType: contentType,
	})
	if err != nil {
		return 0, err
	}
	return info.Size, nil
}

File: services/worker/src/storage.rs
use async_trait::async_trait;
use std::path::PathBuf;

#[async_trait]
pub trait Storage {
    async fn download(&self, key: &str) -> anyhow::Result<PathBuf>;
    async fn upload(&self, key: &str, path: &PathBuf, content_type: &str) -> anyhow::Result<()>;
}

pub struct S3Storage;

impl S3Storage {
    pub async fn new(
        _endpoint: &str,
        _bucket: &str,
        _access_key: &str,
        _secret_key: &str,
    ) -> anyhow::Result<Self> {
        Ok(Self)
    }
}

#[async_trait]
impl Storage for S3Storage {
    async fn download(&self, _key: &str) -> anyhow::Result<PathBuf> {
        Ok(PathBuf::new())
    }
    async fn upload(&self, _key: &str, _path: &PathBuf, _content_type: &str) -> anyhow::Result<()> {
        Ok(())
    }
}
